portfolio/src/main/java/com/google/sps/servlets/MilkDataServlet.java

```diff
@@ -0,0 +1,120 @@
+        Double.parseDouble(cells[8]), Double.parseDouble(cells[9])
+      );
+
+      milkConsumption.add(milkYear);
+    }
+    scanner.close();
+
+    return milkConsumption;
+  }
+
+  private ArrayList<DairyYear> generateRelativeMilkData(ArrayList<DairyYear> originalData) {
+    ArrayList<DairyYear> relativeMilkConsumption = new ArrayList<DairyYear>();
+    
+    DairyYear baseYear = new DairyYear(1975, 0, 0, 0, 0, 0, 0, 0, 0, 0);
+    relativeMilkConsumption.add(baseYear);
+
+    for(int i = 0; i < originalData.size() - 1; i++) {
+      baseYear = originalData.get(i);
+      DairyYear compareYear = originalData.get(i + 1);
+
+      // Calculate the differences
+      double wholeD = calculatePercentageChange(baseYear.getWhole(), compareYear.getWhole());
+      double reducedFatD = calculatePercentageChange(baseYear.getReducedFat(),
+                                                     compareYear.getReducedFat());
+      double lowFatD = calculatePercentageChange(baseYear.getLowFat(), compareYear.getLowFat());
+      double skimD = calculatePercentageChange(baseYear.getSkim(), compareYear.getSkim());
+      double flavoredWholeD = calculatePercentageChange(baseYear.getFlavoredWhole(),
+                                                        compareYear.getFlavoredWhole());
+      double flavoredNonWholeD = calculatePercentageChange(baseYear.getFlavoredNonwhole(),
+                                                           compareYear.getFlavoredNonwhole());
+      double buttermilkD = calculatePercentageChange(baseYear.getButtermilk(),
+                                                     compareYear.getButtermilk());
+      double eggnogD = calculatePercentageChange(baseYear.getEggnog(), compareYear.getEggnog());
+      double totalMilkD = calculatePercentageChange(baseYear.getTotalMilk(),
+                                                    compareYear.getTotalMilk());
+
+      DairyYear relativeYear = new DairyYear(
+        compareYear.getYear(), wholeD, reducedFatD, lowFatD, skimD, flavoredWholeD,
+        flavoredNonWholeD, buttermilkD, eggnogD, totalMilkD
+      );
+      relativeMilkConsumption.add(relativeYear);
+    }
+    return relativeMilkConsumption;
+  }
+
+  /**
+   * Calculate the percentage difference of two doubles in comparison to the original.
+   * @param original The 'base' number to consider in the calculation
+   * @param newData The number to compare to the original
+   * @return The percentage difference between the two numbers
+   */
+  private static double calculatePercentageChange(double original, double newData) {
+    return ((newData - original) / original);
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("application/json");
+    Gson gson = new Gson();
+    HashMap<String, ArrayList<DairyYear>> dairyData = new HashMap<String, ArrayList<DairyYear>>();
```

Comment:
nit: Use `Map<String, List<DairyYear>> dairyData` as the variable type (https://stackoverflow.com/questions/9852831/polymorphism-why-use-list-list-new-arraylist-instead-of-arraylist-list-n).